libcaf_io/src/detail/remote_group_module.cpp

```diff
@@ -0,0 +1,153 @@
+ *                                                                            *
+ * If you did not receive a copy of the license files, see                    *
+ * http://opensource.org/licenses/BSD-3-Clause and                            *
+ * http://www.boost.org/LICENSE_1_0.txt.                                      *
+ ******************************************************************************/
+
+#include "caf/detail/remote_group_module.hpp"
+
+#include "caf/detail/group_tunnel.hpp"
+#include "caf/error.hpp"
+#include "caf/expected.hpp"
+#include "caf/io/middleman.hpp"
+#include "caf/make_counted.hpp"
+#include "caf/sec.hpp"
+
+namespace caf::detail {
+
+remote_group_module::remote_group_module(io::middleman* mm)
+  : super(mm->system(), "remote"), mm_(mm) {
+  // nop
+}
+
+void remote_group_module::stop() {
+  nodes_map tmp;
+  critical_section([this, &tmp] {
+    using std::swap;
+    if (!stopped_) {
+      stopped_ = true;
+      swap(nodes_, tmp);
+    }
+  });
+  for (auto& nodes_kvp : tmp)
+    for (auto& instances_kvp : nodes_kvp.second)
+      instances_kvp.second.get()->stop();
+}
+
+expected<group> remote_group_module::get(const std::string& group_locator) {
+  return mm_->remote_group(group_locator);
+}
+
+group_tunnel_ptr remote_group_module::get_impl(const node_id& origin,
+                                               const std::string& group_name) {
+  CAF_ASSERT(origin != none);
+  bool lazy_connect = false;
+  auto instance = critical_section([&, this] {
+    if (stopped_) {
+      return group_tunnel_ptr{nullptr};
+    } else {
+      auto& instances = nodes_[origin];
+      if (auto i = instances.find(group_name); i != instances.end()) {
+        return i->second;
+      } else {
+        lazy_connect = true;
+        auto instance = make_counted<detail::group_tunnel>(this, group_name,
+                                                           origin);
+        instances.emplace(group_name, instance);
+        return instance;
+      }
+    }
+  });
+  if (lazy_connect)
+    mm_->resolve_remote_group_intermediary(origin, group_name,
+                                           make_callback(instance));
+  return instance;
+}
+
+group_tunnel_ptr remote_group_module::get_impl(actor intermediary,
+                                               const std::string& group_name) {
+  CAF_ASSERT(intermediary != nullptr);
+  return critical_section([&, this]() {
+    if (stopped_) {
+      return group_tunnel_ptr{nullptr};
+    } else {
+      auto& instances = nodes_[intermediary.node()];
+      if (auto i = instances.find(group_name); i != instances.end()) {
+        auto result = i->second;
+        result->connect(std::move(intermediary));
+        return result;
+      } else {
+        auto result = make_counted<detail::group_tunnel>(
+          this, group_name, std::move(intermediary));
+        instances.emplace(group_name, result);
+        return result;
```

Comment:
It would make sense to rename `result` to `instance` which would be consistent with `get_impl(const node_id&, const std::string&)`. This is a nitpick though.